### PR TITLE
Don't rely on use_io_memory default

### PR DESCRIPTION
**Pull request by mandolaerik**
- Avoid legacy Register_LE form
- Enable `use_io_memory` for generated io_memory_access overrides
- Make explicit assumptions on use_io_memory
- Use io_memory explicitly in shared 1.2/1.4 banks

### CODE REVIEW

**Comment by syssimics** [comment]
Verification [#11479201](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11479201): fail

**Comment by syssimics** [comment]
Verification [#11480090](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11480090): fail

**Comment by syssimics** [comment]
Verification [#11480124](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11480124): fail

**Comment by syssimics** [comment]
Verification [#11484978](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11484978): pass

**Comment by mandolaerik** [comment]
[Trigger Jenkins](https://teamcity01-ir.devtools.intel.com/viewQueued.html?itemId=11489489&tab=queuedBuildOverviewTab)

**Comment by mandolaerik** [comment]
jenkins looks ok, errors in cobalt-pvc-6.0,connorsville-6.0,falconshores-6.0 seem unrelated

File: test/1.2/misc/dml14.dml [hunk 1]
```diff
@@ -225,22 +225,147 @@ bank field_regs {
...
+bank io_memory_access_bank is io_memory_access_bank;
+
+bank io_memory_access_bank_compat is (
+    io_memory_access_bank, dml12_compat_io_memory_access) {
+    method test() {
+        default();
+        local uint32_le_t value;
+        local buffer_t buf;
+        buf.len = 4;
+        buf.data = cast(&value, uint8 *);
```
Comment (lwaern-intel) [inline test/1.2/misc/dml14.dml:271]: Why not:
```C
local buffer_t buf = { .len = 4, .data = cast(&value, uint8 *) };
```
?

File: test/1.4/lib/T_io_memory.dml [hunk 2]
```diff
@@ -54,9 +55,10 @@ bank d[i < 2][j < 2] {
 }
 
 bank e {
-    method io_memory_access(generic_transaction_t *memop, uint64 offset,
-                            void *aux) -> (bool) {
-        return default(memop, offset, cast(1234, void *));
+    param use_io_memory = false;
+    method transaction_access(transaction_t *t, uint64 offset,
+                              void *aux) -> (exception_type_t) {
+        return default(t, offset, cast(1234, void *));
```
Comment (lwaern-intel) [inline test/1.4/lib/T_io_memory.dml:61]: Seeing how the test is called `io_memory`, should the `transaction_access` case really be tested here? (This is a 1.4 test, not 1.2 where `param use_io_memory = false` still actually uses `io_memory`)

File: test/1.4/misc/porting-common-compat.dml [hunk 3]
```diff
@@ -198,6 +200,7 @@ bank read_access_memop {
 }
 
 bank write_access_memop {
+    param use_io_memory = true;
```
Comment (lwaern-intel) [inline test/1.4/misc/porting-common-compat.dml:203]: Consider using:
```C
in each bank { param use_io_memory = true; }
```
instead

File: test/1.4/misc/porting-common.dml [hunk 4]
```diff
@@ -150,6 +152,7 @@ bank read_access_memop {
 }
 
 bank write_access_memop {
+    param use_io_memory = true;
```
Comment (lwaern-intel) [inline test/1.4/misc/porting-common.dml:155]: Ditto: consider using:
```C
in each bank { param use_io_memory = true; }
```
instead

File: lib/1.2/dml12-compatibility.dml [hunk 5]
```diff
@@ -63,40 +63,131 @@ template write_only_attr is pseudo_attr {
...
+            // transaction_access will get the memop back untouched
+            // through ATOM_get_transaction_memop
+            list = {
+                ATOM_memop(memop),
+                ATOM_list_end(0),
+                ATOM_list_end(0),
+                ATOM_list_end(0),
+                ATOM_list_end(0),
+                ATOM_list_end(0),
+                ATOM_list_end(0),
```
Comment (lwaern-intel) [inline lib/1.2/dml12-compatibility.dml:128]: This is silly. Do you need every element beyond the first `ATOM_list_end` to have determinate value? Why not:
```C
list[0] = ATOM_memop(memop);
list[1] = ATOM_list_end(0);
```
And leave it at that?

File: lib/1.2/dml12-compatibility.dml [hunk 6]
```diff
@@ -63,40 +63,131 @@ template write_only_attr is pseudo_attr {
...
 }
 
-// If a bank in a DML 1.4 file overrides io_memory_access, then
+// If a bank in a DML 1.4 file overrides io_memory_access
+// or transaction_access, then
 // this template can be instantiated to make the override effective when
 // imported from a DML 1.2 device.
 template dml12_compat_io_memory_access is bank {
-    method access(generic_transaction_t *memop, physical_address_t offset,
+    param _have_dml12_compat_io_memory_access = true;
```
Comment (lwaern-intel) [inline lib/1.2/dml12-compatibility.dml:71]: Unused?

File: RELEASENOTES-1.2.docu [hunk 7]
```diff
@@ -163,4 +163,9 @@
   <build-id value="6215"><add-note> Improve porting rules for method
       return values: In some simple cases, avoid storing the method's
       return value in an intermediate local variable.</add-note></build-id>
+  <build-id value="next"><add-note> If a bank declared with
+      <tt>use_io_memory=false</tt> in a DML 1.4 file is imported from
+      a DML 1.2 device, then the device will now honor overrides of
+      the <tt>transaction_access</tt> method. This simplifies future
```
Comment (lwaern-intel) [inline RELEASENOTES-1.2.docu:169]: Wait, maybe I lost a piece of the thread along the line -- isn't this only true if `dml12_compat_io_memory_access` is instantiated?

**Comment by jhbaarnh** [comment]
Is there a build based on the 7 branch as proof that it works?

**Review by lwaern-intel — CHANGES_REQUESTED** [review]

File: test/1.4/misc/porting-common-compat.dml (hunk 3 above)
Comment (mandolaerik) [reply test/1.4/misc/porting-common-compat.dml:203]: no, these files are generated by port-dml

File: test/1.4/lib/T_io_memory.dml (hunk 2 above)
Comment (mandolaerik) [reply test/1.4/lib/T_io_memory.dml:61]: it's primarily a test of how to implement io_memory on top level, as required by old PCI, and how to make function numbers and stuff work with that. One of the new things is that also transaction-based banks are supported there.

File: RELEASENOTES-1.2.docu (hunk 7 above)
Comment (mandolaerik) [reply RELEASENOTES-1.2.docu:169]: Or if transaction_access is called. Should indeed be rephrased, it was written for an earlier revision of the PR.

File: lib/1.2/dml12-compatibility.dml (hunk 6 above)
Comment (mandolaerik) [reply lib/1.2/dml12-compatibility.dml:71]: thanks! leftover from an earlier revision

File: lib/1.2/dml12-compatibility.dml [hunk 8]
```diff
@@ -63,40 +63,131 @@ template write_only_attr is pseudo_attr {
...
+    method access(generic_transaction_t *mop, physical_address_t offset,
                   physical_address_t size) throws {
-        if (!this.io_memory_access(memop, offset, NULL))
+        if (!this.io_memory_access(mop, offset, NULL))
```
[the remark this answers is not in the document]
Comment (mandolaerik) [reply lib/1.2/dml12-compatibility.dml:74]: It served a purpose in a previous revision. Now it's noise, reverting

File: lib/1.2/dml12-compatibility.dml (hunk 5 above)
Comment (mandolaerik) [reply lib/1.2/dml12-compatibility.dml:128]: good catch

File: lib/1.2/dml12-compatibility.dml [hunk 9]
```diff
@@ -63,40 +63,131 @@ template write_only_attr is pseudo_attr {
...
+
+    method transaction_access(transaction_t *t, uint64 offset,
+                              void *aux) -> (exception_type_t) default {
+        log info, 4: "entering compatibility wrapper"
+            + " for transaction_access overrides";
+        local generic_transaction_t *memop = ATOM_get_transaction_memop(t);
+        local generic_transaction_t mop;
+        local bool is_read = SIM_transaction_is_read(t);
+        local uint64 size = SIM_transaction_size(t);
+        local uint8 buf[size];
```
[the remark this answers is not in the document]
Comment (mandolaerik) [reply lib/1.2/dml12-compatibility.dml:148]: thanks

File: test/1.2/misc/dml14.dml (hunk 1 above)
Comment (mandolaerik) [reply test/1.2/misc/dml14.dml:271]: Thanks, fixed (in a bunch of places)

File: test/1.4/misc/porting-common.dml (hunk 4 above)
Comment (mandolaerik) [reply test/1.4/misc/porting-common.dml:155]: generated by port-dml

**Comment by syssimics** [comment]
Verification [#11501699](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11501699): fail

File: RELEASENOTES-1.2.docu [hunk 10]
```diff
@@ -165,7 +165,11 @@
       return value in an intermediate local variable.</add-note></build-id>
   <build-id value="next"><add-note> If a bank declared with
       <tt>use_io_memory=false</tt> in a DML 1.4 file is imported from
-      a DML 1.2 device, then the device will now honor overrides of
-      the <tt>transaction_access</tt> method. This simplifies future
-      transition to the Simics 7 API.</add-note></build-id>
+      a DML 1.2 device, using the `dml12-compatibility.dml` library,
+      then a call to the <tt>transaction_access</tt> method will now
+      trigger a proper bank access. Also, if a bank uses the
+      `adml12_compat_io_memory_access` template, then overrides of
```
Comment (lwaern-intel) [inline RELEASENOTES-1.2.docu:171]: Use `<tt> </tt>` instead of `` ` ` ``, and `s/adml12_compat_io_memory_access/dml12_compat_io_memory_access`

**Review by lwaern-intel — CHANGES_REQUESTED** [review]

**Comment by syssimics** [comment]
Verification [#11504380](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11504380): pass

**Comment by jhbaarnh** [comment]
@mandolaerik Is there a build based on the 7 branch as proof that it works?

**Comment by mandolaerik** [comment]
> I am mostly concerned that it works in 7. Will approve when I see evidence of that.

Test build: https://github.com/intel-innersource/applications.simulators.simics.simics-base/pull/3861

**Comment by mandolaerik** [comment]
> I am mostly concerned that it works in 7. Will approve when I see evidence of that.

[This one](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11531827&) should suffice?

File: RELEASENOTES-1.2.docu (hunk 10 above)
Comment (mandolaerik) [reply RELEASENOTES-1.2.docu]: Done

File: RELEASENOTES-1.2.docu (hunk 7 above)
Comment (mandolaerik) [reply RELEASENOTES-1.2.docu]: Rephrased.

**Review by jhbaarnh — APPROVED** [review]

File: lib/1.2/dml12-compatibility.dml [hunk 11]
```diff
@@ -75,7 +75,7 @@ template dml12_compat_io_memory_access is bank {
     }
 }
 
-template dml12_compat_access_override {
+in each bank {
     // 'in each':ed in all banks to permit calls to the otherwise DML
     // 1.4-specific functions io_memory_access and transaction_access.
     // Can be instantiated explicitly to avoid rank ambiguities.
```
Comment (lwaern-intel) [inline lib/1.2/dml12-compatibility.dml:81]: Remove this "can be instantiated" line, too

File: lib/1.2/dml12-compatibility.dml [hunk 12]
```diff
@@ -63,40 +63,131 @@ template write_only_attr is pseudo_attr {
     param readable = false;
 }
 
-// If a bank in a DML 1.4 file overrides io_memory_access, then
+// If a bank in a DML 1.4 file overrides io_memory_access
+// or transaction_access, then
 // this template can be instantiated to make the override effective when
 // imported from a DML 1.2 device.
```
[the remark this answers is not in the document]
Comment (mandolaerik) [reply lib/1.2/dml12-compatibility.dml:69]: I suggest to do this as a separate step, https://jira.devtools.intel.com/browse/SIMICS-21052

**Review by lwaern-intel — CHANGES_REQUESTED** [review]
1.2 compatibility displaces me from my happy place

Anyway address that one small requested change and I'll approve this

**Comment by mandolaerik** [comment]
> Verification [#11534643](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11534643): fail

Unrelated intermittent failure

**Comment by syssimics** [comment]
Verification [#11534850](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11534850): pass

**Review by lwaern-intel — APPROVED** [review]

**Pull request merged**